sstables/uncompress.cc

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include <cstdint>
+#include <seastar/core/fstream.hh>
+
+#include "seastar/core/future.hh"
+#include "seastar/core/temporary_buffer.hh"
+
+#include "types.hh"
+#include "exceptions.hh"
+#include "checksum_utils.hh"
+
+namespace sstables {
+
+// File data source implementation for uncompressed SSTables
+template <ChecksumUtils ChecksumType>
+class uncompressed_file_data_source_impl : public data_source_impl {
+    std::optional<input_stream<char>> _input_stream;
+    const checksum& _checksum;
+    uint64_t _underlying_pos;
+    uint64_t _pos;
+    uint64_t _beg_pos;
+    uint64_t _end_pos;
+public:
+    uncompressed_file_data_source_impl(file f, const checksum& checksum,
+    uint64_t pos, size_t len, file_input_stream_options options)
+    : _checksum(checksum)
+    , _pos(pos)
+    , _beg_pos(pos)
+    , _end_pos(pos + len)
+    {
+        // _beg_pos and _end_pos specify positions in the stream.
+        // These are not necessarily aligned on chunk boundaries.
+        // To be able to verify the checksums, we need to translate
+        // them into a range of chunks that contain the given
+        // address range, and open a file input stream to read that
+        // range. The _underlying_pos always points to the current
+        // chunk-aligned position of the file input stream.
+        auto chunk_size = checksum.chunk_size;
+        SCYLLA_ASSERT(chunk_size);
+        auto start = (_beg_pos / chunk_size) * chunk_size;
+        auto end = (_end_pos / chunk_size + 1) * chunk_size;
+        _input_stream = make_file_input_stream(std::move(f), start, end - start, std::move(options));
+        _underlying_pos = start;
+    }
+
+    virtual future<temporary_buffer<char>> get() override {
+        auto chunk_size = _checksum.chunk_size;
+        if (_pos >= _end_pos) {
+            return make_ready_future<temporary_buffer<char>>();
+        }
+        // Read the next chunk. We need to skip part of the first
+        // chunk, but then continue to read from beginning of chunks.
+        // Also, we need to take into account that the last chunk can
+        // be smaller than `chunk_size'.
+        if (_pos != _beg_pos && (_pos % chunk_size) != 0) {
+            throw std::runtime_error("uncompressed reader out of sync");
+        }
+        return _input_stream->read_exactly(chunk_size).then([this, chunk_size](temporary_buffer<char> buf) {
+            auto expected_checksum = _checksum.checksums[_pos / chunk_size];
+            auto actual_checksum = ChecksumType::checksum(buf.get(), buf.size());
+            if (expected_checksum != actual_checksum) {
+                throw sstables::malformed_sstable_exception(format("uncompressed chunk of size {} at file offset {} failed checksum, expected={}, actual={}", buf.size(), _underlying_pos, expected_checksum, actual_checksum));
```

Comment:
Will this error be further decorated by the file name? I expect it will, but need to make sure.

Reply:
I'm afraid not, but we do log the file name in a separate log message. The exception gets caught by `sstable::validate()` which decorates the message with the keyspace and table, and also reports the number of errors to the caller. The caller, `scrub_sstables_validate_mode()`, logs the SSTable as being invalid if any errors were reported.

An example:
```
INFO  2024-08-21 14:10:47,635 [shard  0:main] compaction - Scrubbing in validate mode /tmp/scylla-3fc1a4e7-71e8-4506-ad3f-0148972a118e/me-3giv_0v1z_3klmo2canrtpzud9tc-big-Data.db
ERROR 2024-08-21 14:10:47,635 [shard  0:main] compaction - [Scrub compaction scrub_test_framework.table205] unexpected exception: sstables::malformed_sstable_exception (uncompressed chunk of size 65536 at file offset 0 failed checksum, expected=2816886359, actual=3414968069)
INFO  2024-08-21 14:10:47,635 [shard  0:main] compaction - Finished scrubbing in validate mode /tmp/scylla-3fc1a4e7-71e8-4506-ad3f-0148972a118e/me-3giv_0v1z_3klmo2canrtpzud9tc-big-Data.db - sstable is invalid
```

Reply:
In the regular reader, it will be decorated with sstable name. 

```
--
sstables/mx/reader.cc:            } catch(sstables::malformed_sstable_exception& e) {
sstables/mx/reader.cc:                throw sstables::malformed_sstable_exception(format("Failed to read partition from SSTable {} due to {}", _sst->get_filename(), e.what()));
sstables/mx/reader.cc-            }
```

so if we end up optionally wiring this uncompressed stream based on checksum_chance option, we'll be fine. for the long run, we should consider moving the decoration to a common place, so if we end up implementing a new reader for a newer format, it will not be lost.